laghos_assembly.hpp

```diff
@@ -67,182 +67,111 @@ struct Tensors1D
 {
 private:
-   const int dim, nzones;
-
-   QuadratureData *quad_data;
-   FiniteElementSpace &FESpace;
-   Tensors1D *tensors1D;
-
-   // Mass matrix action on quadrilateral elements in 2D.
-   void MultQuad(const Vector &x, Vector &y) const;
-   // Mass matrix action on hexahedral elements in 3D.
-   void MultHex(const Vector &x, Vector &y) const;
-
+   const MPI_Comm comm;
+   const int dim, NE, vsize;
+   const QuadratureData &qdata;
+   FiniteElementSpace &fes;
+   ParBilinearForm pabf;
+   int ess_tdofs_count;
+   Array<int> ess_tdofs;
+   OperatorPtr mass;
+   Tensors1D *T1D;
 public:
-   MassPAOperator(QuadratureData *quad_data_, FiniteElementSpace &fes,
-                  Tensors1D *t1D)
-      : Operator(fes.GetVSize()),
-        dim(fes.GetMesh()->Dimension()), nzones(fes.GetMesh()->GetNE()),
-        quad_data(quad_data_), FESpace(fes), tensors1D(t1D) { }
-
-   // Mass matrix action.
-   virtual void Mult(const Vector &x, Vector &y) const;
-
-   void ComputeDiagonal2D(Vector &diag) const;
-   void ComputeDiagonal3D(Vector &diag) const;
-
-   virtual const Operator *GetProlongation() const
-   { return FESpace.GetProlongationMatrix(); }
-   virtual const Operator *GetRestriction() const
-   { return FESpace.GetRestrictionMatrix(); }
+   MassPAOperator(Coefficient&, const QuadratureData&,
+                  ParFiniteElementSpace&, const IntegrationRule&, Tensors1D*);
+   virtual void Mult(const Vector&, Vector&) const;
+   virtual void ComputeDiagonal2D(Vector&) const;
+   virtual void ComputeDiagonal3D(Vector&) const;
+   virtual void SetEssentialTrueDofs(Array<int>&);
+   virtual void EliminateRHS(Vector&) const;
 };
 
 // Scales by the inverse diagonal of the MassPAOperator.
 class DiagonalSolver : public Solver
 {
 private:
    Vector diag;
-   FiniteElementSpace &FESpace;
+   FiniteElementSpace &fes;
 
 public:
    DiagonalSolver(FiniteElementSpace &fes)
-      : Solver(fes.GetVSize()), diag(), FESpace(fes) { }
+      : Solver(fes.GetVSize()), diag(), fes(fes) { }
 
    void SetDiagonal(Vector &d)
    {
-      const Operator *P = FESpace.GetProlongationMatrix();
-
-      // Happens when this is called by the serial version of Laghos.
-      if (P == NULL) { diag = d; return; }
-
+      const Operator *P = fes.GetProlongationMatrix();
+      MFEM_VERIFY(P, "No prolongation matrix!");
       diag.SetSize(P->Width());
       P->MultTranspose(d, diag);
    }
 
    virtual void Mult(const Vector &x, Vector &y) const
    {
-      for (int i = 0; i < x.Size(); i++) { y(i) = x(i) / diag(i); }
+      const int N = x.Size();
+      auto d_diag = diag.Read();
+      auto d_x = x.Read();
+      auto d_y = y.Write();
+      MFEM_FORALL(i, N, d_y[i] = d_x[i] / d_diag[i];);
    }
-   virtual void SetOperator(const Operator &op) { }
-};
-
-// Performs partial assembly for the energy mass matrix on a single zone.
-// Used to perform local CG solves, thus avoiding unnecessary communication.
-class LocalMassPAOperator : public Operator
```

Comment:
Is this replaced by something or just removed as a method?

Reply:
It's replaced for now by one global solve for the energy.
We can bring a local one if needed.